### PR TITLE
GitCommitBear.py: Enforce fullmatch on body_regex

### LINKED ISSUE #1331: GitCommitBear/body_regex: Use re.fullmatch instead of re.search?
I'm for using `re.fullmatch` to match the whole body of the commit. Currently you can work with anchors, but with the restriction that they work only on a single line, as `re.MULTILINE` is not activated for `body_regex`. Like for `shortlog_regex`, I would suggest to use `fullmatch` instead of `search`.

CC @nkprince007

### PR DESCRIPTION
**Pull request by nkprince007**
Force body to follow a certain multiline pattern

Fixes #1331

### CODE REVIEW

**Comment by pratyushprakash** [comment]
I don't think this needs 4 commits.

**Comment by nkprince007** [comment]
@pratyushprakash Could you be more specific? Because, the commit message specifies respective changes correctly. 

If you insist in squashing them, please specify the commits that could be squashed.

File: bears/vcs/git/GitCommitBear.py [hunk 1]
```diff
@@ -241,8 +241,9 @@ def check_body(self, body,
                                'HEAD commit. Please add one.')
             return
 
-        if body_regex and not re.search(body_regex, body):
-            yield Result(self, 'No match found in commit message for the '
+        if body_regex and not re.fullmatch(
+                body_regex, body.strip(), re.MULTILINE | re.DOTALL):
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: I wouldn't enable `re.MULTILINE` and `re.DOTALL`, normally people don't expect this + this is not necessary^^

File: bears/vcs/git/GitCommitBear.py [hunk 2]
```diff
@@ -241,8 +241,9 @@ def check_body(self, body,
...
+            yield Result(self, "HEAD commit's body does not match the "
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: messaeg change should go into an own commit ;)

File: tests/vcs/git/GitCommitBearTest.py [hunk 3]
```diff
@@ -271,11 +271,12 @@ def test_body_checks(self):
...
-                        'First line, blablablablablabla.\n'
-                        'Another line, blablablablablabla.\n'
-                        'Fix 1112')
+                        'TICKER---BEGIN\n'
+                        'Bla bla bla\n'
+                        'Fix 1112\n'
+                        'TICKER---END')
         self.assertEqual(self.run_uut(
-                             body_regex=r'Fix\s+[1-9][0-9]*\s*'), [])
+                             body_regex=r'^TICKER---BEGIN.*TICKER---END$'), [])
```
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: you don't need the anchors^^

Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: maybe you want another test case, as you change the example completely?

**Comment by Makman2** [comment]
@pratyushprakash commits must only contain a single change, not 10 ;) That makes review easier and modularizes history :+1: So @nkprince007 does exactly right

**Comment by Makman2** [comment]
ack cf9733a

File: bears/vcs/git/GitCommitBear.py [hunk 4]
```diff
@@ -291,6 +291,9 @@ def check_issue_reference(self, body,
         if host not in self.SUPPORTED_HOST_KEYWORD_REGEX:
             return
 
+        if not re.search(self.SUPPORTED_HOST_KEYWORD_REGEX[host], body):
+            return
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: not necessary imo, the regex below just won't match and you achieve the same effect^^

File: .coafile [hunk 5]
```diff
@@ -31,6 +31,9 @@ bears = SpaceConsistencyBear
 bears = GitCommitBear
 shortlog_trailing_period = False
 shortlog_regex = ([^:]*|[^:]+: [A-Z0-9*].*)
+body_close_issue = True
+body_close_issue_full_url = True
+body_close_issue_last_line = True
```
Comment (Makman2) [inline .coafile]: We don't enforce issue references in our URLs ;)
Like in this commit ;)

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: It won't match, deliberately reporting the result that body doesn't contain any issue reference.

```
README: Modify installation procedure.

Add the details about setting up ``git``
```
The above commit, will be reported false positive for errors, which is unexpected, right?

File: .coafile (hunk 5 above)
Comment (nkprince007) [inline .coafile]: ^^Didn't quite get you. I just followed [this.](https://api.coala.io/en/latest/Developers/Writing_Good_Commits.html)

File: bears/vcs/git/GitCommitBear.py (hunk 2 above)
Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: Sure. :smile:

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: But if you activate `issue_regex` you want to enforce links, don't you?

Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: Yeah, that's the main issue here.
With coala and other repos, we won't have all the commits fixing some bugs, right?
And if we enforce this in `.coafile`, we would require all commits to have issue references.
That is clearly impossible.
So, a check for presence of the keywords tells us that this isn't a commit that fixes/closes something, right? Hope you get the point.

File: bears/vcs/git/GitCommitBear.py (hunk 1 above)
Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: Just multiline, then?

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: actually there are repos which require always an issue reference, or a JIRA ticket reference. So enforcing this on all commits is not completely unuseful.

alright behaviour is okay, but would be good if you implement another flag `body_enforce_issue_reference` to enforce this for all commits. Then you can still keep the "old" behaviour.

Also I think you should modify the code here to implement this behaviour: https://github.com/coala/coala-bears/pull/1333/files/4ce4c848a53d0c9b2024da8b9cea4fedad84ace5#diff-e0a334e6d00dfb6b1fbebb7c515ca4f8R333
In fact you just have to remove this if / extend it by
```python
if body_enforce_issue_reference and if len(results) == 0:
    yield Result(...)
```

File: .coafile (hunk 5 above)
Comment (Makman2) [inline .coafile]: nvm, see above as you change behaviour

File: bears/vcs/git/GitCommitBear.py (hunk 1 above)
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: no flags

File: .coafile (hunk 5 above)
Comment (nkprince007) [inline .coafile]: So, by implementing the above stuff, we could leave these changes as is?

File: tests/vcs/git/GitCommitBearTest.py (hunk 3 above)
Comment (nkprince007) [inline tests/vcs/git/GitCommitBearTest.py]: Actually, I changed this only because the two test cases already in there would be the same, if  I leave the test as is.

File: bears/vcs/git/GitCommitBear.py (hunk 4 above)
Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: Sounds good. But what if a user has a commit like this.
```
Shortlog
First line, blablablablablabla
Fix http://gitlab.com/user/repo/issues/1112
Another line, blablablablablabla.
```
And he has the flags `body_close_issue`, `body_close_issue_full_url` and `body_close_issue_on_last_line` enabled? 
Shouldn't the expected output be something that states there's no issue reference in last line? Instead it yields no result.

File: tests/vcs/git/GitCommitBearTest.py (hunk 3 above)
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: if it's the same like before then you don't need to change it, right? :)

File: bears/vcs/git/GitCommitBear.py [hunk 6]
```diff
@@ -241,7 +241,7 @@ def check_body(self, body,
                                'HEAD commit. Please add one.')
             return
 
-        if body_regex and not re.search(body_regex, body):
+        if body_regex and not re.fullmatch(body_regex, body.strip()):
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: why `strip()`?

File: tests/vcs/git/GitCommitBearTest.py [hunk 7]
```diff
@@ -269,13 +269,15 @@ def test_body_checks(self):
...
+        # body_regex, not fully matched
         self.git_commit('Shortlog\n\n'
                         'First line, blablablablablabla.\n'
                         'Another line, blablablablablabla.\n'
                         'Fix 1112')
         self.assertEqual(self.run_uut(
-                             body_regex=r'Fix\s+[1-9][0-9]*\s*'), [])
+                             body_regex=r'Fix\s+[1-9][0-9]*\s*'),
+                         ['No match found in commit message for the regular '
+                          'expression provided: Fix\s+[1-9][0-9]*\s*'])
```
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: could you perform also a valid fullmatch?

File: .coafile (hunk 5 above)
Comment (Makman2) [inline .coafile]: Could you add `body_enforce_issue_reference = False` explicitly? We shouldn't rely on bear defaults in this case, as this is definitely the setting we want for coala ;) Better explicit than implicit :)

File: bears/vcs/git/GitCommitBear.py (hunk 6 above)
Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: To remove any improbable newlines his/her text editor might be adding. :smile:

File: tests/vcs/git/GitCommitBearTest.py (hunk 7 above)
Comment (nkprince007) [inline tests/vcs/git/GitCommitBearTest.py]: Oops, totally forgot about the case.

File: .coafile (hunk 5 above)
Comment (nkprince007) [inline .coafile]: Yeah. Sure.

File: tests/vcs/git/GitCommitBearTest.py (hunk 7 above)
Comment (Makman2) [inline tests/vcs/git/GitCommitBearTest.py]: ah okay, nvm then ;)

**Comment by Makman2** [comment]
ack 009f46a 2e4e35c 5128b4d 972b979

File: .coafile [hunk 8]
```diff
@@ -31,6 +31,10 @@ bears = SpaceConsistencyBear
...
+body_enforce_issue_reference = False
```
Comment (jayvdb) [inline .coafile]: According to :point_up: [January 26, 2017 5:25 PM](https://gitter.im/coala/coala?at=5889cea3074f7be7631e1e30) , I believe we want this to be `True`.  Please correct me if I have misunderstood.
cc @Makman2.

Comment (nkprince007) [inline .coafile]: That strictly enforces all commits to *have issue reference*
But not all commits do fix/close something I suppose @jayvdb

**Comment by nkprince007** [comment]
IMO, disabling `body_close_issue_last_line` should be enough for the above mentioned stuff.
@jayvdb

File: .coafile (hunk 8 above)
Comment (jayvdb) [inline .coafile]: Right, but as you can see from the Gitter chat, Lasse is happy for all future commits to have an issue created. It is a good practise anyway.

**Comment by jayvdb** [comment]
Yes, we may disable `body_close_issue_last_line` on the release branches as a temporary fix,

File: .coafile (hunk 8 above)
Comment (Makman2) [inline .coafile]: Let's discuss that in a separate issue, for now we should just follow the current convention^^

Comment (nkprince007) [inline .coafile]: @Makman2 @jayvdb please mark the changes, if any, are required.

Else could we get through with the merge?

File: bears/vcs/git/GitCommitBear.py (hunk 6 above)
Comment (jayvdb) [inline bears/vcs/git/GitCommitBear.py]: There should only be a single trailing `\n`.  Please only remove that, instead of `strip()`, which will mean junky git messages can slip through.

Otherwise the users [git config](https://git-scm.com/docs/git-commit#git-commit---cleanupltmodegt) needs fixing, and it should be an error.

**Comment by nkprince007** [comment]
There's a clear chance when someone could actually place something like 
```
Shortlog: amazing shortlog




Fixes gitterHQ/gitter#32
```
What about this one @jayvdb?
https://github.com/coala/coala-bears/pull/1333#discussion_r98097515

**Comment by jayvdb** [comment]
@nkprince007 , that should be an error.  See https://git-scm.com/docs/git-commit#git-commit---cleanupltmodegt , which I mentioned in my previous comment.  It cleans up multiple consecutive blank lines.

**Comment by Makman2** [comment]
@jayvdb @nkprince007 filed https://github.com/coala/coala-bears/issues/1351 for that^^

**Comment by nkprince007** [comment]
Could this be merged then @Makman2 ?

File: bears/vcs/git/GitCommitBear.py (hunk 6 above)
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: @jayvdb I think `strip` is alright, git should guarantee that only a traling \n is there. Also when someone writes commit messages with 10 newlines between body and shortlog, we catch it nicely away. Imo `body.strip()` contains then the real commit body :+1:

**Comment by Makman2** [comment]
ack a8a4418 a25fb0f a2d55e8 bcd0f39

**Comment by nkprince007** [comment]
ack 5302246 3e5c9fa 684befa 41a372c 

**CC** @Makman2

**Comment by Makman2** [comment]
@rultor merge
